README.md
# XmlSerializer

-----

As I tried many different approaches to get a simple to handle XML serializer, I had some issues with all of them (I think, I just didn't find the right one) and started to create my own attempt --- inspired by Microsofts [XmlSerializer](https://msdn.microsoft.com/en-us/library/system.xml.serialization.xmlserializer.aspx?cs-save-lang=1&cs-lang=csharp) class.

## Example

```php
<?php

// include needed class
// only dependency is PHP's own SimpleXML
require_once __DIR__.'/XmlSerializer/src/XmlSerializer.php';

// initialize new XmlSerializer
$serializer = new AMWD\XmlSerializer();

// explicitly allow associative arrays (Dictionary/Hashtable)
$serializer->AllowAssociativeArray = true;

// create test objects
$date = new stdClass();
$date->year = 2015;
$date->month = 9;
$date->day = 25;

$object = new stdClass();
$object->firstname = "Andreas";
$object->lastname = "Mueller";
$object->today = $date;
$object->languages = array('C#', 'PHP', '...');
$object->meetings = array('Breakfast' =>  $date, 'Lunch' => '12:30', 'Dinner' => '19:00');

// serialize test object to XML document
$xml = $serializer->Serialize($object);

// print out XML document of object
echo $xml.PHP_EOL;

// deserialize XML document back to object
$obj = $serializer->Deserialize($xml);

// and print deserialized object
print_r($obj);

?>

```

__Output:__

```
<?xml version="1.0"?>
<stdClass>
  <firstname>Andreas</firstname>
  <lastname>Mueller</lastname>
  <today>
    <year>2015</year>
    <month>9</month>
    <day>25</day>
  </today>
  <languages>C#</languages>
  <languages>PHP</languages>
  <languages>...</languages>
  <meetings>
    <key>Breakfast</key>
    <value>
      <year>2015</year>
      <month>9</month>
      <day>25</day>
    </value>
  </meetings>
  <meetings>
    <key>Lunch</key>
    <value>12:30</value>
  </meetings>
  <meetings>
    <key>Dinner</key>
    <value>19:00</value>
  </meetings>
</stdClass>



SimpleXMLElement Object
(
  [firstname] => Andreas
  [lastname] => Mueller
  [today] => SimpleXMLElement Object
  (
    [year] => 2015
    [month] => 9
    [day] => 25
  )

  [languages] => Array
  (
    [0] => C#
    [1] => PHP
    [2] => ...
  )

  [meetings] => Array
  (
    [0] => SimpleXMLElement Object
    (
      [key] => Breakfast
      [value] => SimpleXMLElement Object
      (
        [year] => 2015
        [month] => 9
        [day] => 25
      )
    )

    [1] => SimpleXMLElement Object
    (
      [key] => Lunch
      [value] => 12:30
    )

    [2] => SimpleXMLElement Object
    (
      [key] => Dinner
      [value] => 19:00
    )
  )
)
```

As you can see, all the magic stuff will be inside the class and you just need to call the simple functions to get it done.

-----

## Changelog

Changelog can be found [here](CHANGELOG)

-----

### LICENSE
My scripts are published under [MIT License](https://am-wd.de/?p=about#license).


CHANGELOG
Please see CHANGELOG at master branch for latest stable version.

2016-07-19
 - 'AllowAssociativeArray' and 'Formatted' are now public properties.
 - AllowAssociativeArray replaces the old setter method.
 - Property 'Formatted' (default: true) indicates a formatted (whitespaces)
   XML after Serialization.
 - Added a method 'RebuildAssociative()' to transform the key-value pairs
   back to an associative array.
 - 'Serialize()' now can get the name of the root node.
 - 'Deserialize()' now keeps the name of the root node.
 - 'GetRootName()' returns the name of the root node from the last deserialization.

2015-11-24
 - Renamed method from 'AssociativeArray()' to 'AllowAssociativeArray()'.

2015-10-28
 - Added UnitTest.
 - Added Makefile for less work.
 - Associative arrays are errors by default.
 - Associative arrays (dictionaries) are difficult to parse.
 - Explicitly enable assoc. arrays with 'AssociativeArray(true)'.

2015-09-25
 - Initial release.
 - very basic functionality.
 - only 'Serialize()' and 'Deserialize()' available.
 - no properties/configuration possible.


tests/XmlSerializerTest.php
<?php
/**
 * XmlSerializerTest.php
 *
 * (c) Andreas Mueller <webmaster@am-wd.de>
 */

namespace AMWD;
require_once __DIR__.'/../src/XmlSerializer.php';

/**
 * Test collection to check functionality of Xml Serializer.
 *
 * @package    AMWD
 * @author     Andreas Mueller <webmaster@am-wd.de>
 * @copyright  (c) 2015 Andreas Mueller
 * @link       https://bitbucket.org/BlackyPanther/xmlserializer
 */
class XmlSerializerTest extends \PHPUnit_Framework_TestCase
{

	public function testSimpleTest()
	{
		$obj = new \stdClass();
		$obj->eins = 'one';
		$obj->zwei = 'two';

		$expected = '<?xml version="1.0"?>'.PHP_EOL.'<stdClass><eins>one</eins><zwei>two</zwei></stdClass>'.PHP_EOL;

		$xml = new XmlSerializer();
		$xml->Formatted = false;
		$serialized = $xml->Serialize($obj);

		$this->assertEquals($expected, $serialized, "Simple Serialization failed.");

		$deserialized = $xml->Deserialize($expected);

		$this->assertEquals($obj->eins, $deserialized->eins, "Simple Deserialize failed on first object.");
		$this->assertEquals($obj->zwei, $deserialized->zwei, "Simple Deserialize failed on second object.");
	}
	
	public function testAdvancedTest()
	{
		$xml = new XmlSerializer();
		$xml->Formatted = false;
		
		$date = new \stdClass();
		$date->year = '2015';
		$date->month = '9';
		$date->day = '25';

		$object = new \stdClass();
		$object->firstname = "Andreas";
		$object->lastname = "Mueller";
		$object->today = $date;
		$object->languages = array('C#', 'PHP', '...');
		$object->meetings = array('Breakfast' =>  $date, 'Lunch' => '12:30', 'Dinner' => '19:00');
		
		$expected = '<?xml version="1.0"?>'.PHP_EOL.'<stdClass><firstname>Andreas</firstname><lastname>Mueller</lastname><today><year>2015</year><month>9</month><day>25</day></today><languages>C#</languages><languages>PHP</languages><languages>...</languages><meetings><key>Breakfast</key><value><year>2015</year><month>9</month><day>25</day></value></meetings><meetings><key>Lunch</key><value>12:30</value></meetings><meetings><key>Dinner</key><value>19:00</value></meetings></stdClass>'.PHP_EOL;
		
		try
		{
			$xml->Serialize($object);
			$this->assertFalse(true, "Advanced Serialize failed => No Exception thrown on disabled assoc. Arrays.");
		}
		catch (\Exception $e)
		{
			// Well done
		}
		
		$xml->AllowAssociativeArray = true;
		$serialized = $xml->Serialize($object);
		
		$this->assertEquals($expected, $serialized, "Advanced Serialization failed.");
		
		$deserialized = $xml->Deserialize($expected);
		
		$this->assertEquals($object->firstname, $deserialized->firstname, "Advanced Deserialization failed on: firstname.");
		$this->assertEquals($object->lastname, $deserialized->lastname, "Advanced Deserialization failed on: lastname.");
		$this->assertEquals($object->today->year, $deserialized->today->year, "Advanced Deserialization failed on: today->year.");
		$this->assertEquals($object->today->month, $deserialized->today->month, "Advanced Deserialization failed on: today->month.");
		$this->assertEquals($object->today->day, $deserialized->today->day, "Advanced Deserialization failed on: today->day.");
		$this->assertEquals($object->languages[0], $deserialized->languages[0], "Advanced Deserialization failed on: languages[0].");
		$this->assertEquals($object->languages[1], $deserialized->languages[1], "Advanced Deserialization failed on: languages[1].");
		$this->assertEquals($object->languages[2], $deserialized->languages[2], "Advanced Deserialization failed on: languages[2].");
		$this->assertEquals($object->meetings["Breakfast"]->year, $deserialized->meetings["Breakfast"]->year, "Advanced Deserialization failed on: meetings[Breakfast]->year.");
		$this->assertEquals($object->meetings["Breakfast"]->month, $deserialized->meetings["Breakfast"]->month, "Advanced Deserialization failed on: meetings[Breakfast]->month.");
		$this->assertEquals($object->meetings["Breakfast"]->day, $deserialized->meetings["Breakfast"]->day, "Advanced Deserialization failed on: meetings[Breakfast]->day.");
		$this->assertEquals($object->meetings["Lunch"], $deserialized->meetings["Lunch"], "Advanced Deserialization failed on: meetings[Lunch].");
		$this->assertEquals($object->meetings["Dinner"], $deserialized->meetings["Dinner"], "Advanced Deserialization failed on: meetings[Dinner].");
	}

}

?>


src/XmlSerializer.php
<?php

/**
 * XmlSerializer.php
 *
 * (c) Andreas Mueller <webmaster@am-wd.de>
 */

namespace AMWD;

class_exists('SimpleXMLElement') || user_error('SimpleXMLElement class is required for XmlSerializer', E_USER_ERROR);
class_exists('DOMDocument') || user_error('DOMDocument class is recommended for XmlSerializer', E_USER_WARNING);

/**
 * Xml serialization for PHP.
 *
 * Xml serialization of objects to string and vice versa.
 * Inspired by Microsofts C# XmlSerializer class.
 * Uses SimpleXML for serialization.
 *
 * @package    AMWD
 * @author     Andreas Mueller <webmaster@am-wd.de>
 * @copyright  (c) 2015 Andreas Mueller
 * @license    MIT - http://am-wd.de/index.php?p=about#license
 * @link       https://bitbucket.org/BlackyPanther/xmlserializer
 * @version    v1.1-20160719 | stable
 */
class XmlSerializer
{

	// --- Fields
	// ===========================================================================

	/**
	 * Data array for all properties
	 * @var array
	 */
	private $data;

	/**
	 * Internal version number
	 * @var string
	 */
	private $version = "1.0";
	
	/**
	 * Remind the name of the root element to write it back on serialization.
	 * @var string
	 */
	private $rootName = "";
	
	/**
	 * Gets or sets a value indicating whether associative arrays are allowed.
	 *
	 * Associative arrays are difficult to parse. Therefore they are not supported in C#
	 * and you need to allow them here explicitly.
	 *
	 * @var bool
	 */
	public $AllowAssociativeArray;
	
	/**
	 * Gets or sets a value indicating whether the serialized output is formatted with whitespaces.
	 * 
	 * @var bool
	 */
	public $Formatted = true;



	// --- 'magic' methods
	// ===========================================================================

	/**
	 * Initializes a new instance of XmlSerializer.
	 *
	 * @return XmlSerializer
	 */
	function __construct()
	{
		$this->AllowAssociativeArray = false;
	}

	/**
	 * 'magic' get method for all properties
	 *
	 * @param string $name name of the property
	 * @return mixed
	 */
	function __get($name)
	{
		if (array_key_exists($name, $this->data))
		{
			return $this->data[$name];
		}

		$trace = debug_backtrace();
		trigger_error('Undefined key for __get(): '
		              .$name.' in '
		              .$trace[0]['file'].' at row '
		              .$trace[0]['line']
		  , E_USER_NOTICE);

		return null;
	}

	/**
	 * 'magic' set method for all properties
	 *
	 * @param string $name name of the property
	 * @param mixed $value value of the property
	 * @return void
	 */
	function __set($name, $value)
	{
		$this->data[$name] = $value;
	}

	/**
	 * 'magic' check if property exists
	 *
	 * @param string $name name of the property
	 * @return bool
	 */
	function __isset($name)
	{
		return isset($this->data[$name]);
	}

	/**
	 * 'magic' property delete
	 *
	 * @param string $name name of the property
	 * @return void
	 */
	function __unset($name)
	{
		if (isset($this->data[$name]))
		{
			unset($this->data[$name]);
		}
	}

	/**
	 * 'magic' override for strings representation
	 *
	 * @return string
	 */
	function __toString()
	{
		return 'XmlSerializer v'.$this->version.' by AM.WD - http://am-wd.de';
	}

	// --- Getter/Setter
	// ===========================================================================
	public function GetRootName()
	{
		return $this->rootName;
	}


	// --- Public Methods
	// ===========================================================================

	/**
	 * Serializes the specified object and returns the XML document as string.
	 *
	 * @param  mixed   $obj       Object to serialialize.
	 * @param  string  $nodename  Name of the root element on serialization. If empty the property 'rootName' is used.
	 *
	 * @return string
	 *
	 * @uses \SimpleXMLElement to create XML structure.
	 * @uses \DOMDocument to format XML document.
	 */
	public function Serialize($obj, $nodename = '')
	{
		if (is_object($obj))
		{
			if (empty($nodename) && empty($this->rootName))
			{
				$root = new \SimpleXMLElement('<'.get_class($obj).'/>');
			}
			else if (empty($nodename) && !empty($this->rootName))
			{
				$root = new \SimpleXMLElement('<'.$this->rootName.'/>');
			}
			else
			{
				$root = new \SimpleXMLElement('<'.$nodename.'/>');
			}
			
			$this->SerializeObject($obj, $root);
		}
		else if (is_array($obj))
		{
			$root = new \SimpleXMLElement('<Array/>');
			$this->SerializeArray($obj, $root);
		}
		else
		{
			$name = gettype($obj);
			$root = new \SimpleXMLElement('<'.$name.'>'.$obj.'</'.$name.'>');
		}

		if ($this->Formatted && class_exists('DOMDocument'))
		{
			$dom = new \DOMDocument("1.0");
			$dom->preserveWhiteSpace = true;
			$dom->formatOutput = true;
			$dom->loadXml($root->asXML());

			return $dom->saveXml();
		}

		return $root->saveXML();
	}

	/**
	 * Deserializes the XML document contained by the string.
	 *
	 * @param string $str Serialized XML document.
	 *
	 * @return mixed
	 */
	public function Deserialize($str)
	{
		libxml_use_internal_errors(true);

		if (!($xml = simplexml_load_string($str)))
		{
			throw new \InvalidArgumentException("XML document seems not to be valid");
		}
		
		preg_match("/<([^?].*)>/", $str, $matches);
		if (!empty($matches[1]))
		{
			$this->rootName = $matches[1];
		}
		
		if ($this->AllowAssociativeArray)
		{
			$xml = $this->RebuildAssociative($xml);
		}
		
		return $xml;
	}

	// --- Protected Methods
	// ===========================================================================

	/**
	 * Checks if given array is associative or iterative.
	 *
	 * @param array $array Array to check.
	 *
	 * @return bool
	 */
	protected static function is_assoc($array)
	{
		$key = key($array);
		return is_array($array) && !is_int($key);
	}

	/**
	 * Serializes an object to XML.
	 *
	 * @param mixed $obj Object to serialize.
	 * @param \SimpleXMLElement $node current xml node to add object to.
	 *
	 * @return void
	 */
	protected function SerializeObject($obj, $node)
	{
		$properties = get_object_vars($obj);

		foreach ($properties as $key => $value)
		{
			if (is_object($value))
			{
				$new = $node->addChild($key);
				$this->SerializeObject($value, $new);
			}
			else if (is_array($value))
			{
				$this->SerializeArray($value, $node, $key);
			}
			else
			{
				$node->addChild($key, $value);
			}
		}
	}

	/**
	 * Serializes an array to XML.
	 *
	 * If given array is associative, the key will be saved as attribute.
	 *
	 * @param mixed[] $array Array to serialize.
	 * @param \SimpleXMLElement $node current xml node to add array to.
	 * @param string $childname name of node to create.
	 *
	 * @return void
	 */
	protected function SerializeArray($array, $node, $childname = null)
	{
		$assoc = self::is_assoc($array);
		
		if ($assoc)
		{
			if (!$this->AllowAssociativeArray)
			{
				throw new \UnexpectedValueException("Associative arrays are not allowed");
			}
			
			foreach ($array as $key => $value)
			{
				$name = ($childname == null) ? gettype($value) : $childname;
				
				if (is_object($value))
				{
					$name = ($childname == null) ? get_class($value) : $childname;
					
					$kvp = $node->addChild($name);
					$kvp->addChild('key', $key);
					$new = $kvp->addChild('value');
					$this->SerializeObject($value, $new);
				}
				else if (is_array($value))
				{
					$kvp = $node->addChild($name);
					$kvp->addChild('key', $key);
					$new = $kvp->addChild('value');
					$this->SerializeArray($value, $new);
				}
				else
				{
					$kvp = $node->addChild($name);
					$kvp->addChild('key', $key);
					$kvp->addChild('value', $value);
				}
			}
		}
		else
		{
			foreach ($array as $key => $value)
			{
				$name = ($childname == null) ? gettype($value) : $childname;
				
				if (is_object($value))
				{
					$name = ($childname == null) ? get_class($value) : $childname;
					
					$new = $node->addChild($name);
					$this->SerializeObject($value, $new);
				}
				else if (is_array($value))
				{
					$new = $node->addChild($name);
					$this->SerializeArray($value, $new);
				}
				else
				{
					$node->addChild($name, $value);
				}
			}
		}
	}
	
	/**
	 * Tries to rebuild the Functionality of associative arrays.
	 * 
	 * @param  mixed  $obj  Input for the function. Can be every datatype.
	 * 
	 * @return mixed
	 */
	protected function RebuildAssociative($obj)
	{
		switch (gettype($obj))
		{
			case 'object':
				$ar = (array)$obj;
				if (count($ar) == 1 && isset($ar[0]))
				{
					return $ar[0];
				}
				else if (isset($obj->key) && isset($obj->value))
				{
					$tmp = array();
					$tmp[$ar['key']] = $ar['value'];
					return $tmp;
				}
				else
				{
					$res = new \stdClass();
					
					foreach ($ar as $key => $value)
					{
						switch (gettype($value))
						{
							case 'array':
								$res->$key = array();
								if (isset($value[0]->key) && isset($value[0]->value))
								{
									foreach ($value as $i => $kvp)
									{
										$tmp = $this->RebuildAssociative($kvp);
										$res->$key[array_keys($tmp)[0]] = $tmp[array_keys($tmp)[0]];
									}
								}
								else
								{
									foreach ($value as $k => $v)
									{
										$res->$key[$k] = $v;
									}
								}
								break;
							default:
								$res->$key = $value;
								break;
						}
					}
					
					/*$ar = (array)$res;
					if (count($ar) == 1)
					{
						$key = array_keys($ar)[0];
						return $res->$key;
					}*/
					
					return $res;
				}
				break;
			default:
				return $obj;
		}
	}
}

?>

example.php
<?php

// include needed class
// only dependency is PHP's own SimpleXML
require_once __DIR__.'/src/XmlSerializer.php';

// initialize new XmlSerializer
$serializer = new AMWD\XmlSerializer();

// explicitly allow associative arrays (Dictionary/Hashtable)
$serializer->AllowAssociativeArray = true;

// create test objects
$date = new stdClass();
$date->year = 2015;
$date->month = 9;
$date->day = 25;

$object = new stdClass();
$object->firstname = "Andreas";
$object->lastname = "Mueller";
$object->today = $date;
$object->languages = array('C#', 'PHP', '...');
$object->meetings = array('Breakfast' =>  $date, 'Lunch' => '12:30', 'Dinner' => '19:00');

// serialize test object to XML document
// flag indicates function to format whitespaces before return
$xml = $serializer->Serialize($object);

// print out XML document of object
echo $xml.PHP_EOL;

// deserialize XML document back to object
$obj = $serializer->Deserialize($xml);

// and print deserialized object
print_r($obj);

?>